File: src/021/problem021.rb
#!/usr/bin/env ruby
class Integer
  def factors
    (1..Math.sqrt(self)).select { |x| (self % x).zero? }.each_with_object([]) do |number, list|
      list << number
      list << self / number unless self == number || number == 1
    end
  end

  def amicable?
    sum_of_factors = factors.reduce(:+)
    return false if self == sum_of_factors
    self == sum_of_factors.factors.reduce(:+)
  end
end

puts (1..10_000).find_all(&:amicable?).reduce(:+)

File: src/021/problem021.py
#!/usr/bin/env python
from functools import reduce

def factors(n, remove_own=False):
    f = set(reduce(list.__add__, ([i, n//i] for i in range(1, int(pow(n, 0.5) + 1)) if n % i == 0)))
    if remove_own:
        f.remove(n)
    return f

def amicable(n):
    dn = sum(factors(n, True))
    return dn > 0 and sum(factors(dn, True)) == n and dn != n

print(sum([n for n in range(1, 10000) if amicable(n)]))
